flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java

```diff
@@ -0,0 +1,334 @@
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * The implementation of {@link FlinkKubeClient}.
+ */
+public class Fabric8FlinkKubeClient implements FlinkKubeClient {
+
+	private static final Logger LOG = LoggerFactory.getLogger(Fabric8FlinkKubeClient.class);
+
+	private final Configuration flinkConfig;
+
+	private final KubernetesClient internalClient;
+
+	private final List<Decorator<ConfigMap, FlinkConfigMap>> configMapDecorators;
+
+	private final List<Decorator<Service, FlinkService>> internalServiceDecorators;
+	private final List<Decorator<Service, FlinkService>> restServiceDecorators;
+
+	private final List<Decorator<Deployment, FlinkDeployment>> flinkMasterDeploymentDecorators;
+
+	private final List<Decorator<Pod, FlinkPod>> taskManagerPodDecorators;
+
+	@Nonnull
+	private final String clusterId;
+
+	@Nonnull
+	private final String nameSpace;
+
+	public Fabric8FlinkKubeClient(Configuration flinkConfig, KubernetesClient client) {
+		this.internalClient = client;
+		this.flinkConfig = flinkConfig;
+		this.configMapDecorators = new ArrayList<>();
+		this.internalServiceDecorators = new ArrayList<>();
+		this.restServiceDecorators = new ArrayList<>();
+		this.flinkMasterDeploymentDecorators = new ArrayList<>();
+		this.taskManagerPodDecorators = new ArrayList<>();
+
+		this.clusterId = flinkConfig.getString(KubernetesConfigOptions.CLUSTER_ID);
+		this.nameSpace = flinkConfig.getString(KubernetesConfigOptions.NAMESPACE);
+
+		initialize();
+	}
+
+	private void initialize() {
+		this.configMapDecorators.add(new InitializerDecorator<>(Constants.CONFIG_MAP_PREFIX + clusterId));
+		this.configMapDecorators.add(new OwnerReferenceDecorator<>());
+		this.configMapDecorators.add(new ConfigMapDecorator());
+
+		this.internalServiceDecorators.add(new InitializerDecorator<>(clusterId));
+		this.internalServiceDecorators.add(new ServiceDecorator(
+			KubernetesConfigOptions.ServiceExposedType.ClusterIP, false));
+
+		this.restServiceDecorators.add(new InitializerDecorator<>(clusterId + Constants.FLINK_REST_SERVICE_SUFFIX));
+		String exposedType = flinkConfig.getString(KubernetesConfigOptions.REST_SERVICE_EXPOSED_TYPE);
+		this.restServiceDecorators.add(new ServiceDecorator(
+			KubernetesConfigOptions.ServiceExposedType.valueOf(exposedType), true));
+		this.restServiceDecorators.add(new OwnerReferenceDecorator<>());
+
+		this.flinkMasterDeploymentDecorators.add(new InitializerDecorator<>(clusterId, Constants.APPS_API_VERSION));
```

Comment:
Sorry but I'm not sure why the master needs a different API version `Constants.APPS_API_VERSION` here? Is it related to `Constants.API_VERSION` but with a `apps/` prefix, or can be a totally independent version?

Reply:
In kubernetes api specification, each resource will have a stable api version. By default, the resources `Pod` `Service` `ConfigMap` are basic resource type. It has been included in `v1`. The `Deployment` is an additional resource. It exists in `apps/v1`. They are different api scope even the only difference is `apps`. Maybe in the future, the stable basic api version is `v1.1` and stable additional api version is still `apps/v1`.

Reply:
I see. Thanks for the explanation.